lucene/core/src/java/org/apache/lucene/index/DefaultIndexingChain.java

```diff
@@ -527,45 +589,63 @@ private void indexPoint(PerField fp, IndexableField field) throws IOException {
+          }
+          return DocValues.emptyNumeric();
+        }
 
-          case BINARY:
-            throw new IllegalArgumentException("invalid doc value type:" + dvType + " for sortField:" + sortField);
+        @Override
+        public BinaryDocValues getBinaryDocValues(String field) {
+          if (Objects.equals(field, fieldToValidate) && dvType != DocValuesType.BINARY) {
+            throw new IllegalArgumentException("SortField " + sortField + " expected field [" + field + "] to be BINARY but it is [" + dvType + "]");
+          }
+          return DocValues.emptyBinary();
+        }
 
-          case SORTED:
-            if (sortField.getType().equals(SortField.Type.STRING) == false) {
-              throw new IllegalArgumentException("invalid doc value type:" + dvType + " for sortField:" + sortField);
-            }
-            break;
+        @Override
+        public SortedDocValues getSortedDocValues(String field) {
+          if (Objects.equals(field, fieldToValidate) && dvType != DocValuesType.SORTED) {
+            throw new IllegalArgumentException("SortField " + sortField + " expected field [" + field + "] to be SORTED but it is [" + dvType + "]");
+          }
+          return DocValues.emptySorted();
+        }
 
-          case SORTED_NUMERIC:
-            if (sortField instanceof SortedNumericSortField == false) {
-              throw new IllegalArgumentException("invalid doc value type:" + dvType + " for sortField:" + sortField);
-            }
-            break;
+        @Override
+        public SortedNumericDocValues getSortedNumericDocValues(String field) {
+          if (Objects.equals(field, fieldToValidate) && dvType != DocValuesType.SORTED_NUMERIC) {
+            throw new IllegalArgumentException("SortField " + sortField + " expected field [" + field + "] to be SORTED_NUMERIC but it is [" + dvType + "]");
+          }
+          return DocValues.emptySortedNumeric(0);
+        }
 
-          case SORTED_SET:
-            if (sortField instanceof SortedSetSortField == false) {
-              throw new IllegalArgumentException("invalid doc value type:" + dvType + " for sortField:" + sortField);
-            }
-            break;
+        @Override
+        public SortedSetDocValues getSortedSetDocValues(String field) {
+          if (Objects.equals(field, fieldToValidate) && dvType != DocValuesType.SORTED_SET) {
+            throw new IllegalArgumentException("SortField " + sortField + " expected field [" + field + "] to be SORTED_SET but it is [" + dvType + "]");
+          }
+          return DocValues.emptySortedSet();
+        }
 
-          default:
-            throw new IllegalArgumentException("invalid doc value type:" + dvType + " for sortField:" + sortField);
+        @Override
+        public FieldInfos getFieldInfos() {
+          throw new UnsupportedOperationException();
         }
-        break;
-      }
+
+        @Override
+        public int maxDoc() {
+          return 0;
```

Comment:
I'd have a slight preferenrece for throwing an UOE here, would it work?

Reply:
`IndexSorter.getDocComparator(Reader)` calls `maxDoc()` on the reader to allocate its comparison arrays.  We could change the signature so that it takes `maxDoc` as well as the Reader, which would reduce the indirection a bit and make it easier to read?

Reply:
+1